kernel/src/hil/hash.rs

```diff
@@ -0,0 +1,227 @@
+//! In kernel abstraction over hash functions
+//!
+//! This is an abstraction over hash functions and implementations of
+//! these hash functions. It supports exotic hash functions and
+//! checksums with non-byte outputs and inputs.
+//!
+//! The constraints over the Hasher are enforced by the HashType type
+//! parameter.
+//!
+//! Authors
+//! -------------------
+//! * Leon Schuermann <leon.git@is.currently.online>
+//! * Daniel Rutz <info@danielrutz.com>
+//! * April 01, 2020
+
+use crate::ReturnCode;
+
+/// A trait to be implemented on zero-sized types indicating
+/// properties and types of hash functions
+///
+/// Implementors should be zero sized
+pub trait HashType {
+    /// The input type of the hash function
+    ///
+    /// Typically a byte
+    type Input = u8;
```

Comment:
hmm, is this really worth the effort of introducing the unstable `associated type defaults`?

Also, I don't get why input is needed at all, can you explain?

Reply:
I'm fine either way. I didn't want to clutter the initial proposal. I think this is a detail which makes the code cleaner, but ultimately doesn't add value.